observability/src/main/java/io/grpc/observability/LoggingChannelProvider.java

```diff
@@ -0,0 +1,97 @@
+/*
+ * Copyright 2022 The gRPC Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc.observability;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import com.google.common.annotations.VisibleForTesting;
+import io.grpc.ChannelCredentials;
+import io.grpc.ExperimentalApi;
+import io.grpc.ManagedChannelBuilder;
+import io.grpc.ManagedChannelProvider;
+import io.grpc.ManagedChannelRegistry;
+import io.grpc.netty.NettyChannelProvider;
+import io.grpc.observability.interceptors.LoggingChannelInterceptor;
+
+/** A channel provider that injects logging interceptor. */
+@ExperimentalApi("https://github.com/grpc/grpc-java/issues/8869")
+final class LoggingChannelProvider extends ManagedChannelProvider {
+  @VisibleForTesting NettyChannelProvider prevProvider;
+
+  @VisibleForTesting static LoggingChannelProvider instance;
+
+  private LoggingChannelProvider() {
+    ManagedChannelProvider provider = ManagedChannelProvider.provider();
+    if (provider instanceof NettyChannelProvider) {
+      prevProvider = (NettyChannelProvider) provider;
+    } else {
+      throw new ProviderNotFoundException("Only NettyChannelProvider supported!");
```

Comment:
Any reason it has to be `NettyChannelProvider`? Does the impl need to use any API from grpc-netty?

Reply:
Good question. Originally I didn't want to rely on `NettyChannelProvider`. But in `ManagedChannelProvider` the methods `builderForAddress` and `builderForTarget` etc are protected whereas they are public in `NettyChannelProvider` so are callable from here. If that anomaly is resolved I can remove the dependency on `NettyChannelProvider`. Unless you have a better suggestion...

Reply:
This will not work for `netty-shaded`, `AltsChannelBuilder`, `ComputeEngineChannelBuilder` etc.. The `protected` accessor problem can be resolved as follows:

```java
  @Override
  protected ManagedChannelBuilder<?> builderForAddress(String name, int port) {
    return addInterceptor(ManagedChannelBuilder.forAddress(name, port));
  }

  @Override
  protected ManagedChannelBuilder<?> builderForTarget(String target) {
    return addInterceptor(ManagedChannelBuilder.forTarget(target));
  }

  @Override
  public NewChannelBuilderResult newChannelBuilder(String target, ChannelCredentials creds) {
    ManagedChannelBuilder<?> builder = Grpc.newChannelBuilder(target, creds);
    if (builder != null) {
      return NewChannelBuilderResult.channelBuilder(
              addInterceptor(builder));
    }
    NewChannelBuilderResult result = NewChannelBuilderResult.channelBuilder(builder);
    checkNotNull(result.getError(), "Expected error to be set!");
    return result;
  }
```

Reply:
Oh, the workaround can not work because this provider will be immediately registered in `init()`.

Reply:
Yes, calling `ManagedChannelBuilder.forAddress` will result into infinite recursion. It is possible to avoid recursion by detecting it thru ThreadLocal variables etc. but looks hacky. Let me see how to make it work for netty-shaded...

Reply:
Fixed as per our discussion offline. PTAL